src/tribler-core/tribler_core/components/tag/db/tag_db.py

```diff
@@ -191,10 +191,31 @@ def show_suggestions_condition(torrent_tag):
         return self._get_tags(infohash, show_suggestions_condition)
 
     def get_infohashes(self, tags: Set[str]) -> List[bytes]:
-        """Get list of infohashes that belongs to the tag. Only tags with condition `_show_condition` will be returned
+        """Get list of infohashes that belongs to the tag.
+        Only tags with condition `_show_condition` will be returned.
+        In the case that the tags set contains more than one tag,
+        only torrents that contain all `tags` will be returned.
         """
-        return select(tt.torrent.infohash for tt in self.instance.TorrentTag
-                      if self._show_condition(tt) and tt.tag.name in tags).fetch()
+
+        # first, get all torrents that contains any tag from `tags`
```

Comment:
**issue:** The result of a query that returns torrents with *any* tags may be several orders larger than the result of a query that returns torrents with *all* tags. Retrieving all these objects into memory may be very slow.

**suggestion:** It is better to select torrents with *all* tags directly in the SQL query.

SQL does not have a direct way to express the FORALL universal quantifier. To express the necessary condition, it is necessary to use a double negation: "select all torrents for which does not exist a tag from the list of required tags that is not associated with this torrent". It will be something like:

```python
# torrents that contains all necessary tags
query_results = select(
    torrent for torrent in self.instance.Torrent
    if not exists(
        tag for tag in self.instance.Tag
        if tag.name in tags and not exists(
            torrent_tag for torrent_tag in self.instance.TorrentTag
            if torrent_tag.torrent == torrent
            and torrent_tag.tag == tag
            and self._show_condition(torrent_tag)
        )
    )
)
```

Queries with double `not exists` may be hard to read and understand at first, but they are the proper way to express the FORALL quantifier in SQL.

Reply:
@kozlovsky wow, looks like black magic 👏 
Thanks. Implemented.